src/led/libled.h

```diff
@@ -0,0 +1,492 @@
+/*
+ * LED library public interface
+ *
+ * Copyright (C) 2022-2022 Red Hat, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#ifndef _LIB_LED_INCLUDED_
+#define _LIB_LED_INCLUDED_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <linux/limits.h>
+#include <stdbool.h>
+
+#define LED_SYM_PUBLIC __attribute__((visibility("default")))
+
+typedef int led_status_t;
+
+struct led_ctx;
+
+enum led_log_level_enum {
+	LED_LOG_LEVEL_UNDEF = 0,
+	LED_LOG_LEVEL_QUIET,
+	LED_LOG_LEVEL_ERROR,
+	LED_LOG_LEVEL_WARNING,
+	LED_LOG_LEVEL_INFO,
+	LED_LOG_LEVEL_DEBUG,
+	LED_LOG_LEVEL_ALL,
+};
+
+enum led_cntrl_type {
+	LED_CNTRL_TYPE_UNKNOWN = 0,
+	LED_CNTRL_TYPE_DELLSSD,
+	LED_CNTRL_TYPE_VMD,
+	LED_CNTRL_TYPE_SCSI,
+	LED_CNTRL_TYPE_AHCI,
+	LED_CNTRL_TYPE_NPEM,
+	LED_CNTRL_TYPE_AMD,
+};
+
+/**
+ * @brief The following functions are for functionality pertaining to hardware controllers.
+ */
+
+/**
+ * @brief Translates a controller string ID to enumerated type
+ * 
+ * @param[in]	cntrl_str	String representation of controller type 
+ * @return enum led_cntrl_type 
+ */
+enum led_cntrl_type LED_SYM_PUBLIC led_string_to_cntrl_type(const char *cntrl_str);
+
+/**
+ * @brief Translates the enumerated controller type to string representation
+ * 
+ * @param[in]	cntrl		Enumerated constant 
+ * @return string representation of controller
+ * 
+ * Note: Do not free string, it is a library constant
+ */
+const LED_SYM_PUBLIC char  *const led_cntrl_type_to_string(enum led_cntrl_type cntrl);
+
+/**
+ * @brief Opaque data type for controller list entry
+ */
+struct led_cntrl_list_entry;
+
+/**
+ * @brief Opaque data type for controller list
+ */
+struct led_cntrl_list;
```

Comment:
I'm not exactly sure what you're saying here, please elaborate a bit more .  I do agree having a function to reset the iterator would be useful.

Reply:
Here we have a family of functions to manipulate internal lib lists (we are cloning list to be safe):
```C
led_status_t LED_SYM_PUBLIC led_cntrls_get(struct led_ctx *ctx, struct led_cntrl_list **cntrls);
void LED_SYM_PUBLIC led_cntrl_list_free(struct led_cntrl_list *cntrls);
LED_SYM_PUBLIC struct led_cntrl_list_entry *led_cntrl_next(struct led_cntrl_list *cntrls);
```
I think that we can simplify that too:
```C
// Get next list element
led_status_t  led_cntrl_get_next(struct led_cntrl_list *cntrls)
// Reset current cntrl list entry pointer
led_status_t  led_cntrl_list_reset(void)
```
We are initializing `cntrl_devices` inside `ctx`. All we need is to provide particular list element.
Inside `ctx` or `sysfs` we can place `curr_cntrl_devices_list_ptr` and just modify it when `_next` function is called.
I think that it is safe because of the rule mentioned above:
> Always use a library context object. Every thread should then
> operate on its own context.

So, as a result, we will just need a opaque type for list entry only.

Reply:
Done